consensus/service/src/byzantine_ledger/worker.rs

```diff
@@ -340,6 +298,89 @@ impl<
+        }
+
+        // Sync service reports that we are out of sync.
+        match &self.ledger_sync_state {
+            LedgerSyncState::InSync => LedgerSyncState::MaybeBehind(now),
+            LedgerSyncState::MaybeBehind(behind_since) => {
+                let is_behind_duration = now - *behind_since;
+                if is_behind_duration > IS_BEHIND_GRACE_PERIOD {
+                    LedgerSyncState::IsBehind {
+                        attempt_sync_at: now,
+                        num_sync_attempts: 0,
+                    }
+                } else {
+                    // No change, keep waiting.
+                    LedgerSyncState::MaybeBehind(*behind_since)
+                }
+            }
+            LedgerSyncState::IsBehind {
+                attempt_sync_at,
+                num_sync_attempts,
+            } => {
+                // No change, still behind.
+                LedgerSyncState::IsBehind {
+                    attempt_sync_at: *attempt_sync_at,
+                    num_sync_attempts: *num_sync_attempts,
+                }
+            }
+        }
+    }
+
+    // Incrementally sync the ledger.
+    fn sync_next_blocks(&mut self, num_blocks: u32) {
+        let num_sync_attempts = if let LedgerSyncState::IsBehind {
+            num_sync_attempts, ..
+        } = &self.ledger_sync_state
+        {
+            *num_sync_attempts
+        } else {
+            panic!("Attempted to sync when not behind?");
+        };
+
+        self.ledger_sync_state = match self
+            .ledger_sync_service
+            .attempt_ledger_sync(&self.network_state, num_blocks)
+        {
+            Ok(()) => {
+                // Synced a chunk of blocks, but may still be behind.
+                LedgerSyncState::IsBehind {
+                    attempt_sync_at: Instant::now(),
+                    num_sync_attempts: 0,
+                }
+            }
+            Err(err) => {
+                // Reattempt with capped linear backoff.
+                log::warn!(self.logger, "Could not sync ledger: {:?}", err);
+                let next_sync_at =
+                    Instant::now() + Duration::from_secs(min(num_sync_attempts + 1, 60));
+                LedgerSyncState::IsBehind {
+                    attempt_sync_at: next_sync_at,
+                    num_sync_attempts: num_sync_attempts + 1,
+                }
+            }
+        };
+    }
+
+    /// Clear any pending values that are no longer be valid.
+    fn update_pending_values(&mut self) {
+        let tx_manager = self.tx_manager.clone();
+        self.pending_values_map
+            .retain(|tx_hash, _| tx_manager.validate(tx_hash).is_ok());
+        // (Help the borrow checker)
+        let self_pending_values_map = &self.pending_values_map;
+        self.pending_values
+            .retain(|tx_hash| self_pending_values_map.contains_key(tx_hash));
+
+        debug_assert!(self.pending_values_map.len() == self.pending_values.len());
```

Comment:
I think this should be a regular assert. We rarely stress this code in debug builds. This applies to other asserts in this file.